process/blocksPool.go

```diff
@@ -90,71 +111,80 @@ func (bp *blocksPool) pruneStorer(index uint64) error {
 }
 
 // PutBlock will put the provided outport block data to the pool
-func (bp *blocksPool) PutBlock(hash []byte, outportBlock *outport.OutportBlock, index uint64) error {
-	shardID := outportBlock.ShardID
-
-	outportBlockBytes, err := bp.marshaller.Marshal(outportBlock)
-	if err != nil {
-		return err
-	}
-
+func (bp *blocksPool) PutBlock(hash []byte, value []byte, newIndex uint64, shardID uint32) error {
 	bp.mutMap.Lock()
 	defer bp.mutMap.Unlock()
 
-	currentIndex, ok := bp.indexesMap[shardID]
+	previousIndex, ok := bp.previousIndexesMap[shardID]
 	if !ok {
-		bp.indexesMap[shardID] = initIndex
-		currentIndex = initIndex
+		bp.previousIndexesMap[shardID] = initIndex
+		previousIndex = initIndex
 	}
 
-	if currentIndex == initIndex {
-		err := bp.storer.Put(hash, outportBlockBytes)
+	if previousIndex == initIndex {
+		err := bp.storer.Put(hash, value)
 		if err != nil {
 			return err
 		}
 
-		bp.indexesMap[shardID] = index
+		bp.previousIndexesMap[shardID] = newIndex
 
 		return nil
 	}
 
-	metaIndex := bp.indexesMap[core.MetachainShardId]
+	isSuccesiveIndex := previousIndex+1 == newIndex
+	if !isSuccesiveIndex {
+		return fmt.Errorf("%w: new index should succesive, previous index %d, new index %d",
+			ErrFailedToPutBlockDataToPool, previousIndex, newIndex)
+	}
 
-	if !bp.shouldPutBlockData(currentIndex, metaIndex) {
+	if !bp.shouldPutBlockData(previousIndex) {
 		return ErrFailedToPutBlockDataToPool
 	}
 
-	err = bp.storer.Put(hash, outportBlockBytes)
+	err := bp.storer.Put(hash, value)
 	if err != nil {
 		return err
 	}
 
-	bp.indexesMap[shardID] = index
+	bp.previousIndexesMap[shardID] = newIndex
 
 	return nil
 }
 
-func (bp *blocksPool) shouldPutBlockData(index, baseIndex uint64) bool {
+func (bp *blocksPool) shouldPutBlockData(index uint64) bool {
+	baseIndex := bp.previousIndexesMap[core.MetachainShardId]
+
 	diff := float64(int64(index) - int64(baseIndex))
 	delta := math.Abs(diff)
 
 	return math.Abs(delta) < float64(bp.maxDelta)
 }
 
-// GetBlock will return outport block data from the pool
-func (bp *blocksPool) GetBlock(hash []byte) (*outport.OutportBlock, error) {
-	marshalledData, err := bp.storer.Get(hash)
+func (bp *blocksPool) setCheckpoint(checkpoint *data.BlockCheckpoint) error {
+	checkpointBytes, err := bp.marshaller.Marshal(checkpoint)
```

Comment:
as you may have noticed so far, I am a big fan of chaining error messages. I believe it is easier to find a defined error message `failed to unmarshal BlockCheckpoint: %w', rather than the error returned by `json.Marshal' since you cannot search the codebase by it.